### ISSUE TITLE
Assert at app exit when opening corrupted image

### ISSUE THREAD

**Issue by riviera-kid** [OpenImageIO/oiio#1817]
### Problem

This simple program shows assert at exit:

`#include <OpenImageIO/imageio.h>`
`int main()`
`{`
`  OIIO::ImageInput *in = OIIO::ImageInput::open("obrtest.png");`
`  //_fcloseall();`
`  return 0;`
`}`

obrtest.png:

![obrtest](https://user-images.githubusercontent.com/7337247/33382495-2de90a5e-d521-11e7-8787-47a2d5061b9e.png)

The problem is probably related to Windows - Visual Studio - zlib. 
If the file is invalid, then all plugins are tried (including zfile plugin).
It seems that there can be a problem in zfile.cpp with closing file handles in zlib and msvc.

The included program shows me in debug mode an assert at app exit (or calling _fcloseall manually):
Expression: (_osfile(fh) & FOPEN)

And it also crashes in release mode on some machines.

When I exclude the zfile plugin from the OIIO lib, it works - but there could be a better solution...

### Versions

* OIIO branch/version: 1.8.5 (static lib), zlib (static) - all compiled with /MD (/MDd in debug)
* OS: Windows 10 x64
* C++ compiler: Visual Studio 2017 x64

**Comment by riviera-kid** [comment]
Seems that the problem is related to MSVC /MT /MD mismatch in my compilation of OIIO

**Comment by lgritz** [comment]
I think this was fixed with #1839. Can you confirm?

**Comment by riviera-kid** [comment]
It works (1.8.9)

**Comment by riviera-kid** [comment]
Just tested it in debug mode, at it still shows assert :(

**Comment by riviera-kid** [comment]
J've just tested 1.8.8 and it is still crashing :(

As a temporary workaround, I've put "return false" to 
ZfileInput::valid_file (const std::string &filename) const
ZfileInput::open (const std::string &name, ImageSpec &newspec)
(I don't need a Z file format...)

**Comment by lgritz** [comment]
Which assert is it hitting? Is it an OIIO assert?

What source file and line does the assert implicate?

**Comment by riviera-kid** [comment]
I've debug the issue and it seems that there could be a problem in zlib <-> MSVC. 
```
FILE *f = fopen("obrtest.png", "rb");
gzFile gz = gzdopen(fileno(f), "rb");
gzclose(gz);
_fcloseall(); // this shows the assert:
/*
File: minkernel\crts\ucrt\src\appcrt\lowio\close.cpp
Line: 49

Expression: (_osfile(fh) & FOPEN)
*/
```
The _fcloseall func is called automatically when the app exits.

This works ok (without an assert):
```
FILE *f = fopen("obrtest.png", "rb");
gzFile gz = gzdopen(fileno(f), "rb");
fclose(f);
_fcloseall();
```

**Comment by lgritz** [comment]
Ah, I see. We are using a different open/close call sequence for open() versus valid_file().

**Comment by lgritz** [comment]
We could switch to the other sequence -- fopen(filename), then gzdopen(fileno(f)), for the valid_file (versus doing a direct gzopen(filename). And see if that helps?

**Comment by lgritz** [comment]
Do you want me to take a stab at a patch to address this?

**Comment by riviera-kid** [comment]
I think that closing the file with gzclose should be valid - according to zlib docs, but it is not working in MSVC (?).
Using directly gzopen(filename)/gzclose should be working. Currently I don't need a patch, but it could be fixed in the next release. Of course if you provide a patch, I can test it for you...
Thanks

**Comment by lgritz** [comment]
Proposed fix in #2070.

@riviera-kid, I've never seen the crash, so I'm going to need to rely on you to try it out on your end with the debug build and see if it appears to clear up the problem.
